File: app/src/main/java/com/savvasdalkitsis/gameframe/infra/navigation/ApplicationNavigator.java
package com.savvasdalkitsis.gameframe.infra.navigation;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;

import com.savvasdalkitsis.gameframe.GameFrameApplication;
import com.savvasdalkitsis.gameframe.infra.TopActivityProvider;
import com.savvasdalkitsis.gameframe.ip.view.IpSetupActivity;

public class ApplicationNavigator implements Navigator {

    private final GameFrameApplication application;
    private TopActivityProvider topActivityProvider;

    public ApplicationNavigator(TopActivityProvider topActivityProvider, GameFrameApplication application) {
        this.topActivityProvider = topActivityProvider;
        this.application = application;
    }

    @Override
    public void navigateToIpSetup() {
        start(createIntent(IpSetupActivity.class));
    }

    private void start(Intent intent) {
        getContext().startActivity(intent);
    }

    private Intent createIntent(Class<?> cls) {
        Context context = getContext();
        Intent intent = new Intent(context, cls);
        if (!(context instanceof Activity)) {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        }
        return intent;
    }

    private Context getContext() {
        Activity topActivity = topActivityProvider.getTopActivity();
        if (topActivity != null) {
            return topActivity;
        }
        return application;
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/navigation/ApplicationNavigator.kt
package com.savvasdalkitsis.gameframe.navigation

import android.app.Activity
import android.content.Context
import android.content.Intent

import com.savvasdalkitsis.gameframe.GameFrameApplication
import com.savvasdalkitsis.gameframe.infra.TopActivityProvider
import com.savvasdalkitsis.gameframe.ip.view.IpSetupActivity

class ApplicationNavigator(private val topActivityProvider: TopActivityProvider, private val application: GameFrameApplication) : Navigator {

    override fun navigateToIpSetup() {
        context.startActivity(createIntent(IpSetupActivity::class.java))
    }

    private fun createIntent(cls: Class<*>): Intent {
        val context = context
        val intent = Intent(context, cls)
        if (context !is Activity) {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        return intent
    }

    private val context: Context
        get() = topActivityProvider.topActivity ?: application
}
